presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/serde/JtsGeometrySerde.java

```
@@ -170,25 +176,26 @@ private static Geometry readPolygon(SliceInput input, boolean multitype)
         LinearRing shell = null;
         List<LinearRing> holes = new ArrayList<>();
         List<Polygon> polygons = new ArrayList<>();
-        for (int i = 0; i < partCount; i++) {
-            Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
-            if (isClockwise(coordinates)) {
-                // next polygon has started
-                if (shell != null) {
-                    polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
-                    holes.clear();
+        try {
+            for (int i = 0; i < partCount; i++) {
+                Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
+                if (isClockwise(coordinates)) {
+                    // next polygon has started
+                    if (shell != null) {
+                        polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+                        holes.clear();
+                    }
+                    shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
                 }
                 else {
-                    verify(holes.isEmpty(), "shell is null but holes found");
+                    holes.add(GEOMETRY_FACTORY.createLinearRing(coordinates));
                 }
-                shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
-            }
-            else {
-                verify(shell != null, "shell is null but hole found");
-                holes.add(GEOMETRY_FACTORY.createLinearRing(coordinates));
             }
+            polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+        }
+        catch (IllegalArgumentException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)